### PR TITLE
(PDB-3328) Munge package inventory facts in terminus

### PR DESCRIPTION
**Pull request by mullr**
This adds terminus support for that package inventory feature. When the
_puppet_inventory_1 fact is seen it is removed, reformatted to be in the tabular
format expected by puppetdb, and placed in a top-level 'package_inventory' key
in the facts command payload.

### CODE REVIEW

**Comment by mullr** [comment]
@puppetlabs-jenkins retest this please

File: puppet/lib/puppet/indirector/facts/puppetdb.rb [hunk 1]
```diff
@@ -26,7 +26,18 @@ def save(request)
         facts = request.instance.dup
         facts.values = facts.values.dup
         facts.values[:trusted] = get_trusted_info(request.node)
-        {
+
+        inventory = facts.values['_puppet_inventory_1']
+        package_inventory =
+          if inventory and inventory.respond_to?(:keys) and inventory['packages']
+            inventory['packages'].map { |p|
+              [p['package_name'], p['version'], p['provider']]
```
Comment (wkalt) [inline puppet/lib/puppet/indirector/facts/puppetdb.rb]: not too big a deal, but seeing as this fact is controlled by us, why not just receive it as a list of tuples and skip this step?

Comment (wkalt) [inline puppet/lib/puppet/indirector/facts/puppetdb.rb]: if we could do that we could also split out the array-valued versions in the fact definition and not have to worry about any of that in the terminus or application

Comment (mullr) [inline puppet/lib/puppet/indirector/facts/puppetdb.rb]: We talked about this; the reason we have maps in the fact but tuples in the command is to keep the more 'public' part of the data (i.e. the data generated on the agent, and sent from  the agent to the master) self describing. We should still be able to split out the array values in the fact definition - I asked Scott Walker to do just that in a comment on the design doc.

**Comment by wkalt** [comment]
@mullr needs a rebase

File: test/puppetlabs/puppetdb/integration/inventory/package_inventory.clj [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      (is (= 0 (count (int/pql-query pdb "facts { name = '_puppet_inventory_1' }")))))
+
+    (comment
+      (testing "Package inventory can be queried"
+        (is (= [{:package_name "openssl"
+                 :version "1.0.2g-1ubuntu4.6"
+                 :provider "apt"}]
+               (int/pql-query pdb "packages { certname = 'agent-with-packages' }")))))))
+
+
```
Comment (wkalt) [inline test/puppetlabs/puppetdb/integration/inventory/package_inventory.clj]: extra linez

File: puppet/lib/puppet/indirector/facts/puppetdb.rb [hunk 3]
```diff
@@ -26,7 +26,15 @@ def save(request)
         facts = request.instance.dup
         facts.values = facts.values.dup
         facts.values[:trusted] = get_trusted_info(request.node)
-        {
+
+        inventory = facts.values['_puppet_inventory_1']
+        package_inventory = nil
+        if inventory && inventory.respond_to?(:keys)
+          package_inventory = inventory['packages']
+          facts.values.delete('_puppet_inventory_1')
```
Comment (wkalt) [inline puppet/lib/puppet/indirector/facts/puppetdb.rb]: we should always be deleting this if it exists -- respond_to? keys is irrelevant

File: test/puppetlabs/puppetdb/integration/inventory/package_inventory.clj [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+                       :env {"FACTER__puppet_inventory_1"
+                             (json/generate-string inventory-fact-value)}}))
+
+    (testing "Ensure that facts were stored"
+      (is (pos? (count (int/pql-query pdb "facts { }")))))
+
+    (testing "Ensure the inventory fact was removed"
+      (is (= 0 (count (int/pql-query pdb "facts { name = '_puppet_inventory_1' }")))))
+
+    (comment
```
Comment (mullr) [inline test/puppetlabs/puppetdb/integration/inventory/package_inventory.clj]: ;; no!

File: puppet/lib/puppet/indirector/facts/puppetdb.rb [hunk 5]
```diff
@@ -26,7 +26,15 @@ def save(request)
         facts = request.instance.dup
         facts.values = facts.values.dup
         facts.values[:trusted] = get_trusted_info(request.node)
-        {
+
+        inventory = facts.values['_puppet_inventory_1']
+        package_inventory = nil
+        if inventory
```
Comment (wkalt) [inline puppet/lib/puppet/indirector/facts/puppetdb.rb]: if _puppet_inventory_1 is null, we still want to delete it I think. Does this guard against that?

**Pull request merged**